xdsl/interpreter.py

```diff
@@ -0,0 +1,273 @@
+            raise InterpretationError(
+                f'Could not find interpretation function for op {op.name}')
+        ft, impl = self._impl_dict[type(op)]
+        return impl(ft, interpreter, op, args)
+
+
+@dataclass
+class InterpreterContext:
+    """
+    Class holding the Python values associated with SSAValues during an interpretation
+    context. An environment is a stack of scopes, values are assigned to the current
+    scope, but can be fetched from a parent scope.
+    """
+
+    name: str = field(default="unknown")
+    parent: InterpreterContext | None = None
+    env: dict[SSAValue, Any] = field(default_factory=dict)
+
+    def __getitem__(self, key: SSAValue) -> Any:
+        """
+        Fetch key from environment. Attempts to first fetch from current scope, then
+        from parent scopes. Raises Interpretation error if not found.
+        """
+        if key in self.env:
+            return self.env[key]
+        if self.parent is not None:
+            return self.parent[key]
+        raise InterpretationError(f'Could not find value for {key} in {self}')
+
+    def __setitem__(self, key: SSAValue, value: Any):
+        """
+        Assign key to current scope. Raises InterpretationError if key already assigned to.
+        """
+        if key in self.env:
+            raise InterpretationError(
+                f'Attempting to register SSAValue {value} for name {key}'
+                f', but value with that name already exists in {self}')
+        self.env[key] = value
+
+    def stack(self) -> Generator[InterpreterContext, None, None]:
+        """
+        Iterates through scopes starting with the root scope.
+        """
+        if self.parent is not None:
+            yield from self.parent.stack()
+        yield self
+
+    def __format__(self, __format_spec: str) -> str:
+        return '/'.join(c.name for c in self.stack())
+
+
+@dataclass
+class Interpreter:
+    """
+    An extensible interpreter, initialised with a Module to intperpret. The implementation
+    for each Operation subclass should be provided via a `InterpretationFunctionTable`
+    instance. Interpretations can be overridden, and the override must be specified
+    explicitly, by passing `override=True` to the `register_functions` method.
+    """
+
+    module: ModuleOp
+    _impls: _InterpreterFunctionImpls = field(
+        default_factory=_InterpreterFunctionImpls)
+    _ctx: InterpreterContext = field(
+        default_factory=lambda: InterpreterContext(name='root'))
+    file: IO[str] | None = field(default=None)
+
+    def get_values(self, values: Iterable[SSAValue]) -> tuple[Any, ...]:
+        """
+        Get values from current environment.
+        """
+        return tuple(self._ctx[value] for value in values)
+
+    def set_values(self, pairs: Iterable[tuple[SSAValue, Any]]):
+        """
+        Set values to current scope.
+        Raises InterpretationError if len(ssa_values) != len(result_values), or if
```

Comment:
Is this documentation still up to date? It seems that the length are always equal?